### PR TITLE
LightSource + Player Refactor

### PR DESCRIPTION
**Pull request by alisterdev**

### CODE REVIEW

File: Assets/Scripts/Light/LightSource.cs [hunk 1]
```diff
@@ -1,102 +1,105 @@
...
+    [Header("Light Source")]
+    [SerializeField]
     [Tooltip("If true, this GameObject can absorb other GameObjects with a LightSource component")]
-    public bool canAbsorb;
+    private bool canAbsorb = false;
     
+    [SerializeField]
     [Tooltip("If true, the player can always absorb this GameObject, even if it has higher light.")]
-    public bool playerWillAlwaysAbsorb;
+    private bool playerWillAlwaysAbsorb = false;
```
Comment (stellarxo) [inline Assets/Scripts/Light/LightSource.cs]: assuming this is for debugging and could probably be removed?

File: Assets/Scripts/Light/LightSource.cs [hunk 2]
```diff
@@ -1,102 +1,105 @@
...
-
-    // DO NOT ACCESS DIRECTLY. Use LightEnergy property instead.
+    private Neighbourhood absorbableLightDetector;
+    
     private LightEnergy lightEnergy;
 
-    public virtual void Awake()
-    {
-    }
+    public virtual void Awake() {}
```
Comment (stellarxo) [inline Assets/Scripts/Light/LightSource.cs]: remove empty functions?

File: Assets/Scripts/Light/LightSource.cs [hunk 3]
```diff
@@ -1,102 +1,105 @@
...
     }
-
+    
+    /// <summary>
+    /// Called the instant the light depletes to zero. 
+    /// Called from the LightEnergy.LightDepleted event.
+    /// </summary>
+    protected virtual void OnLightDepleted()
+    {
+    }
```
Comment (stellarxo) [inline Assets/Scripts/Light/LightSource.cs]: ?? is this overwritten in other classes?

File: Assets/Scripts/Light/LightSource.cs [hunk 4]
```diff
@@ -1,102 +1,105 @@
...
+    {
+        this.LightEnergy.LightDepleted -= OnLightDepleted;
+    }
+        
     /// <summary>
     /// Returns true if this LightSource can absorb the given light source.
+    /// Calculated based on which LightSource has more energy
     /// </summary>
-    public bool CanAbsorb(LightSource otherLightSource)
+    /// <param name="otherLightSource"></param>
```
Comment (stellarxo) [inline Assets/Scripts/Light/LightSource.cs]: I thought we weren't doing params?????

File: Assets/Scripts/Light/LightSource.cs (hunk 2 above)
Comment (alisterdev) [inline Assets/Scripts/Light/LightSource.cs]: @stellarxo  can't remove it, its gonna break player, fish, and every other class that extends LightSource and has override, so merge that, and I'll change all classes one shot on master I guess

File: Assets/Scripts/Light/LightSource.cs (hunk 3 above)
Comment (alisterdev) [inline Assets/Scripts/Light/LightSource.cs]: @stellarxo  Player.cs line 136 I believe

File: Assets/Scripts/Light/LightSource.cs (hunk 4 above)
Comment (alisterdev) [inline Assets/Scripts/Light/LightSource.cs]: They are auto-generated so I do them, but I guess we can clean that for the sake of time

**Pull request merged**